src/Symfony/Component/DependencyInjection/Compiler/CheckArgumentsValidityPass.php

```diff
@@ -0,0 +1,45 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\DependencyInjection\Compiler;
+
+use Symfony\Component\DependencyInjection\Definition;
+use Symfony\Component\DependencyInjection\Exception\RuntimeException;
+
+/**
+ * Checks if arguments of methods are properly configured.
+ *
+ * @author Kévin Dunglas <dunglas@gmail.com>
+ * @author Nicolas Grekas <p@tchwork.com>
+ */
+class CheckArgumentsValidityPass extends AbstractRecursivePass
+{
+    /**
+     * {@inheritdoc}
+     */
+    protected function processValue($value, $isRoot = false)
+    {
+        if (!$value instanceof Definition) {
+            return parent::processValue($value, $isRoot);
+        }
+
+        $arguments = $value->getArguments();
+        if ($arguments !== array_values($arguments)) {
+            throw new RuntimeException(sprintf('Arguments\'s keys of the "%s" service must be integers or names of constructor\'s parameters prefixed by the "$" symbol. Check your service definition.', $this->currentId));
```

Comment:
Exactly the situation where numbering has a hole. The message makes the assumption that if we got there, it might be because of a named arg, even if be don't really know anymore as you spotted.
Which message would be best do you think ?